src/euler_one_qubit_decomposer.rs

```diff
@@ -10,15 +10,636 @@
+            circuit.push((String::from("u1"), vec![tot]));
+        }
+    } else if (theta - PI / 2.).abs() < atol {
+        circuit.push((String::from("u2"), vec![mod_2pi(phi), mod_2pi(lam)]));
+    } else {
+        circuit.push((String::from("u3"), vec![theta, mod_2pi(phi), mod_2pi(lam)]));
+    }
+    OneQubitGateSequence {
+        gates: circuit,
+        global_phase: phase,
+    }
+}
+
+fn circuit_u(
+    theta: f64,
+    phi: f64,
+    lam: f64,
+    phase: f64,
+    simplify: bool,
+    atol: Option<f64>,
+) -> OneQubitGateSequence {
+    let mut circuit = Vec::new();
+    let mut atol = match atol {
+        Some(atol) => atol,
+        None => DEFAULT_ATOL,
+    };
+    if !simplify {
+        atol = -1.0;
+    }
+    let phi = mod_2pi(phi);
+    let lam = mod_2pi(lam);
+    if theta.abs() > atol || phi.abs() > atol || lam.abs() > atol {
+        circuit.push((String::from("u"), vec![theta, phi, lam]));
+    }
+    OneQubitGateSequence {
+        gates: circuit,
+        global_phase: phase,
+    }
+}
+
+fn circuit_psx_gen<F, P, X>(
+    theta: f64,
+    phi: f64,
+    lam: f64,
+    phase: f64,
+    simplify: bool,
+    atol: Option<f64>,
+    mut pfun: P,
+    mut xfun: F,
+    xpifun: Option<X>,
+) -> OneQubitGateSequence
+where
+    F: FnMut(&mut OneQubitGateSequence),
+    P: FnMut(&mut OneQubitGateSequence, f64),
+    X: FnMut(&mut OneQubitGateSequence),
+{
+    let mut phi = phi;
+    let mut lam = lam;
+    let mut theta = theta;
+    let mut circuit = OneQubitGateSequence {
+        gates: Vec::new(),
+        global_phase: phase,
+    };
+    let mut atol = match atol {
+        Some(atol) => atol,
+        None => DEFAULT_ATOL,
+    };
+    if !simplify {
+        atol = -1.0;
+    }
+    // Early return for zero SX decomposition
+    if theta.abs() < atol {
+        pfun(&mut circuit, lam + phi);
+        return circuit;
+    }
+    // Early return for single SX decomposition
+    if (theta - PI / 2.).abs() < atol {
+        pfun(&mut circuit, lam - PI / 2.);
+        xfun(&mut circuit);
+        pfun(&mut circuit, phi + PI / 2.);
+        return circuit;
+    }
+    // General double SX decomposition
+    if (theta - PI).abs() < atol {
+        circuit.global_phase += lam;
+        phi -= lam;
+        lam = 0.;
+    }
+    if mod_2pi(lam + PI).abs() < atol || mod_2pi(phi).abs() < atol {
+        lam += PI;
+        theta = -theta;
+        phi += PI;
+        circuit.global_phase -= theta;
+    }
+    // Shift theta and phi to turn the decomposition from
+    // RZ(phi).RY(theta).RZ(lam) = RZ(phi).RX(-pi/2).RZ(theta).RX(pi/2).RZ(lam)
+    // into RZ(phi+pi).SX.RZ(theta+pi).SX.RZ(lam).
+    theta += PI;
+    phi += PI;
+    circuit.global_phase -= PI / 2.;
+    // emit circuit
+    pfun(&mut circuit, lam);
+    if xpifun.is_some() && mod_2pi(theta).abs() < atol {
+        xpifun.unwrap()(&mut circuit);
+    } else {
+        xfun(&mut circuit);
+        pfun(&mut circuit, theta);
+        xfun(&mut circuit);
+    }
```

Comment:
We could probably keep clippy happy here and avoid suppressing `unnecessary_unwrap` with something like
```rust
match xpifun {
    Some(xpifun) if mod_2pi(theta).abs() < atol => xpifun(&mut circuit),
    _ => {
        xfun(&mut circuit);
        pfun(&mut circuit, theta);
        xfun(&mut circuit);
    }
}
```
though I'd agree that with `if let Some(x) = x && cond` still unstable, it's possibly a bit premature for clippy to be warning on that.

Reply:
Yeah, I bounced off that API not being stabilized and just decided to suppress clippy. I'll try playing with that syntax a bit locally and see if I can get it working as expected

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/pull/9578/commits/b19244405a03268094d94beba1be566750ea40eb

Reply:
Do you have a link to a discussion of "API not being stabilized" ?

Reply:
The tracking issue for the rfc is here: https://github.com/rust-lang/rust/issues/53667 rust upstream hasn't finalized the feature yet. Although looking at the issue and rfc it probably won't be up for stabilization until the next rust edition.

Also, if you try using this syntax the compiler links to this issue too:

```rust
error[[E0658]](https://doc.rust-lang.org/stable/error_codes/E0658.html): `let` expressions in this position are unstable
 --> src/main.rs:5:16
  |
5 |     if true && let Some(foo) = test {
  |                ^^^^^^^^^^^^^^^^^^^^
  |
  = note: [see issue #53667 <https://github.com/rust-lang/rust/issues/53667>](https://github.com/rust-lang/rust/issues/53667) for more information

For more information about this error, try `rustc --explain E0658`.
```